gslib/wildcard_iterator.py

```diff
@@ -643,6 +649,15 @@ def __iter__(self, bucket_listing_fields=None):
     for filepath in filepaths:
       expanded_url = StorageUrlFromString(filepath)
       try:
+        if self.exclude_pattern:
+          base_url_str, _ = os.path.split(wildcard)
```

Comment:
What if the wildcard look something like `f*/b*` that matches `foo/bar`. Won't the base_url_str be incorrect? I am guessing this won't be an issue for rsync as it looks like rsync already expands the wildcard before passing down the string to the WildcardIterator (it calls wildcardIterator twice), but given that this is generic code used for non-rsync commands as well, I think we should address it.

Reply:
So `wildcard` is actually made up of `base_url_str` and the suffix I split off here if you follow the logic back to `rsync.py`, but I do see your point. One way to side-step this entirely is to require an `exclusion_tuple` that contains the original `exclusion_pattern` and the original `base_url_str`, so that way we're not trying to reconstitute it using `wildcard` in some dubious way.

Reply:
Alternatively, we can restrict the exclusion pattern to work only if the `base_url_str` does not have a wildcard, i.e we say that if you want to filter out any patterns, you should not pass in a wildcard string other than the one with format `<wildcard-less path>/**`.  Given that rsync already runs a wildcard iterator on the source to determine that the [source is a single directory](https://github.com/GoogleCloudPlatform/gsutil/blob/bf9394b01bb29d33653e93aad1cd143532d905d6/gslib/commands/rsync.py#L1648), and given the fact that the -x option is currently used only for rsync, it shouldn't be an issue.

It would look something like this
```
base_url_str, _ = os.path.split(wildcard)
if storage_url.WILDCARD_REGEX.match(base_url_str):
  # should never happen for rsync,
  # but just to protect any other command accidentally using this feature in the future.
  raise Error('-x pattern is only allowed for non-wildcard paths')
...
```

For the bug that lead to this fix, one other solution could be to just expand the `continue-on-error` to work for wildcard iterator as well. But this has greater side effect, although we might end up fixing another issue if we do that https://github.com/GoogleCloudPlatform/gsutil/issues/353. Having said that, I think we should still go ahead with fixing -x.

Reply:
Thinking a bit more about this, I think my above proposal is not the best as it does make the logic only relevant to the cases where source is presented as a directory. It will work because rsync always deals with directory, but since wildcard iterator is a generic module, I feel it's not the best. 

So an alternative could be to pass in the exclude pattern with the base_url_str appended. This can be done in the rsync code. So for `gsutil rsync -x "data./.*\.txt$" dir gs://my-bucket`, the exclude pattern will be `dir/data./.*\.txt$` instead of `data./.*\.txt$`. So the entire logic of extracting `base_url_str` can stay in `rsync` and the code in wildcard iterator can stay generic in true sense. I think the logic can go somewhere here https://github.com/GoogleCloudPlatform/gsutil/blob/bf9394b01bb29d33653e93aad1cd143532d905d6/gslib/commands/rsync.py#L734 .

Reply:
I have a version of this completed, I'll be adding it here in a second.